### ISSUE TITLE
Running command in container

### ISSUE THREAD

**Issue by Enkelmann** [fkie-cad/cwe_checker#313]
### Discussed in https://github.com/fkie-cad/cwe_checker/discussions/312

<div type='discussions-op-text'>

<sup>Originally posted by **StepanGavrilov** April  5, 2022</sup>
I have error like

`docker run --rm -v /path_here/hdat2cd_lite_74.iso:/hdat2cd_lite_74.iso fkiecad/cwe_checker hdat2cd_lite_74.iso`

```
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
Execution of Ghidra plugin failed: Process was terminated.
```

```
ERROR REPORT: Import failed for file: /hdat2cd_lite_74.iso (HeadlessAnalyzer)
```

```
ERROR The AutoImporter could not successfully load /hdat2cd_lite_74.iso with the provided import parameters. Please ensure that any specified processor/cspec arguments are compatible with the loader that is used during import and try again. (HeadlessAnalyzer) 
```
</div>

**Comment by Enkelmann** [comment]
Because of the file ending `.iso` of your file I would assume that your file is not a normal binary executable and Ghidra (and thus also the cwe_checker) simply does not know how to handle such a file. You could use the `--bare-metal-config` command line parameter if your file is a bare-metal firmware binary, but then you would also need to provide a configuration file specifying the instruction set architecture that the binary is using. I any case, if your input file is a binary executable I would need some more information about its file format to help you.

The warning at the beginning puzzles me a little. It sounds like your host platform is ARM-based, while the cwe_checker Docker image is compiled for x86-64. I actually never tried whether this would work at all, so I may misinterpret the warning. Anyway, the warning probably has nothing to do with your problem here.

**Comment by StepanGavrilov** [comment]
If i use example in readme, have this trace

"""
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error("missing field `register_calling_convention`", line: 5220, column: 1)', src/caller/src/main.rs:406:64
stack backtrace:
   0:       0x40003507ec - std::backtrace_rs::backtrace::libunwind::trace::hf6a6dfd7da937cb0
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/../../backtrace/src/backtrace/libunwind.rs:90:5
   1:       0x40003507ec - std::backtrace_rs::backtrace::trace_unsynchronized::hc596a19e4891f7f3
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:       0x40003507ec - std::sys_common::backtrace::_print_fmt::hb16700db31584325
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/sys_common/backtrace.rs:67:5
   3:       0x40003507ec - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h231c4190cfa75162
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/sys_common/backtrace.rs:46:22
   4:       0x40003756ec - core::fmt::write::h2a1462b5f8eea807
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/fmt/mod.rs:1163:17
   5:       0x400034ca45 - std::io::Write::write_fmt::h71ddfebc68685972
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/io/mod.rs:1696:15
   6:       0x4000352370 - std::sys_common::backtrace::_print::hcc197d4bebf2b369
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/sys_common/backtrace.rs:49:5
   7:       0x4000352370 - std::sys_common::backtrace::print::h335a66af06738c7c
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/sys_common/backtrace.rs:36:9
   8:       0x4000352370 - std::panicking::default_hook::{{closure}}::h6fac9ac9c8b79e52
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:210:50
   9:       0x4000351f25 - std::panicking::default_hook::h341c1030c6a1161b
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:227:9
  10:       0x4000352a24 - std::panicking::rust_panic_with_hook::h50680ff4b44510c6
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:624:17
  11:       0x4000352500 - std::panicking::begin_panic_handler::{{closure}}::h9371c0fbb1e8465a
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:521:13
  12:       0x4000350c94 - std::sys_common::backtrace::__rust_end_short_backtrace::h9b3efa22a5768c0f
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/sys_common/backtrace.rs:139:18
  13:       0x4000352469 - rust_begin_unwind
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:517:5
  14:       0x400005ed01 - core::panicking::panic_fmt::h23b9203e89cc61cf
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:100:14
  15:       0x400005edf3 - core::result::unwrap_failed::h32ef6b3156e8fc57
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/result.rs:1616:5
  16:       0x4000091ffe - cwe_checker::main::he00131cba5af4cd8
  17:       0x40000992f3 - std::sys_common::backtrace::__rust_begin_short_backtrace::h3f55152a53afb033
  18:       0x40000ad5f9 - std::rt::lang_start::{{closure}}::hc8ee6ccb96ae7b7e
  19:       0x40003504cb - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::hc56adab7a77ec6e3
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/ops/function.rs:259:13
  20:       0x40003504cb - std::panicking::try::do_call::h29f013120c5abc65
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:403:40
  21:       0x40003504cb - std::panicking::try::h86d5b2b66caec4cf
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:367:19
  22:       0x40003504cb - std::panic::catch_unwind::h7dd136d787f51397
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panic.rs:133:14
  23:       0x40003504cb - std::rt::lang_start_internal::{{closure}}::h4a199351e630a8a5
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/rt.rs:128:48
  24:       0x40003504cb - std::panicking::try::do_call::h20ceb9e5dff838c6
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:403:40
  25:       0x40003504cb - std::panicking::try::hc2abb46a5e41bd43
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:367:19
  26:       0x40003504cb - std::panic::catch_unwind::h52c3eb4408ad6dfb
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panic.rs:133:14
  27:       0x40003504cb - std::rt::lang_start_internal::hd15a47be08101c28
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/rt.rs:128:20
  28:       0x4000092d62 - main
  29:       0x4001e88d0a - __libc_start_main
  30:       0x400005f4da - _start
  31:                0x0 - <unknown>

"""

**Comment by StepanGavrilov** [comment]
@Enkelmann Now i changed to base

`#include <stdlib.h>

int main(){
    void* p = malloc(200);
    free(p);
    free(p);
    return 0;
}
`
like in readme and run `gcc example.c -o example-bin`

**Comment by Enkelmann** [comment]
Ah, now I understand. Now this is indeed puzzling. Since I do not expect gcc generating anything strange for such a simple example, my guess would be that it has something to do with your environment. And the only hint I have is that warning at the beginning. Have you tried running the cwe_checker natively instead of using the Docker image?

**Comment by StepanGavrilov** [comment]
@Enkelmann I tested it only in docker. (I use mac with M1)

**Comment by Enkelmann** [comment]
That explains the warning! Ok, my guess is that the emulation of x86 code that is apparently used for executing the Docker image introduces a bug somewhere. But I do not have access to a Mac at the moment, so I cannot test and debug this theory myself. A native installation might help you here, but I also haven't tested whether a native installation of the cwe_checker on a Mac actually works.

**Comment by StepanGavrilov** [comment]
@Enkelmann ok I will try

**Comment by StepanGavrilov** [comment]
@Enkelmann check logs, find it:


```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error("missing field `register_calling_convention`", line: 5220, column: 1)', src/caller/src/main.rs:406:64

```

**Comment by Enkelmann** [comment]
Ok, so it does not work. This either a bug in Ghidra (unlikely, you could check it by running Ghidra separately) or, more likely, a bug in the way that the cwe_checker tries to communicate with Ghidra. And my guess is that the same bug also somehow affects the Docker version of the cwe_checker on your PC. Unfortunately, since I do not own a Mac, I cannot fix it myself at the moment. The only other workaround I can think of right now is to natively install the cwe_checker inside of a virtual machine running Linux.

In conclusion, for the time being Linux will be the only fully supported platform for the cwe_checker. I hope to change that at some point, but the time I can work on stuff like that is unfortunately quite limited.

**Comment by StepanGavrilov** [comment]
@Enkelmann hello again, now i am trying on

```
CPU op-mode(s):        32-bit, 64-bit
Address sizes:         45 bits physical, 48 bits virtual
CPU(s):                  12
Model name:            Intel(R) Xeon(R) Silver 4116 CPU @ 2.10GHz
```

Same error:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error("missing field `register_calling_convention`", line: 5220, column: 1)', src/caller/src/main.rs:406:64
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
But with amd64 no problem.

**Comment by Enkelmann** [comment]
Is this still on MacOS? If yes, then I would be quite confused why it would work on AMD but not on an Intel CPU. At some point in the future we will change the way of communication between Ghidra and the cwe_checker so that it works on operating systems other than Linux. But as I already mentioned, it will unfortunately take some time until I can work on this.

**Comment by StepanGavrilov** [comment]
@Enkelmann this on Ubuntu with intel

**Comment by Enkelmann** [comment]
@StepanGavrilov We now deploy multi-architecture Docker images to Dockerhub (see #321, #322), so if you update the `fkiecad/cwe_checker:latest` image it should run natively on your Mac. Could you test out whether this resolves your problems?

**Comment by Enkelmann** [comment]
Ok, then this is probably some issue with the cwe_checker itself, which I can probably only debug if I find an example binary somewhere where this error occurs.

**Comment by DubcekUnderPineappleTree** [comment]
Most respect to Mr. Enkelmann:
Excuse me,could you please tell me whether you have cwe_checker - master/SRC/ghidra p_code_extractor/PcodeExtractor. Java code analysis documentation?If so,could you please release it to me?Thanks a million!

**Comment by drharryhe** [comment]
@Enkelmann  I run into this issue also. 
Error: missing field `register_calling_convention` at line 57710568 column 1 

Here is the env:
●CPU: Apple  M1 Pro
●ProductName: macOS                                                                                                                                                                                                         
●ProductVersion: 13.4.1                                                                                                                                                                                                        
●BuildVersion:  22F82

Any plan to fix this issue please?

**Comment by Enkelmann** [comment]
@drharryhe  Can you provide the `helloworld` binary? Then I could try to reproduce the issue on my end. But as long as I cannot reproduce it, I won't be able to identify the underlying bug.

**Comment by drharryhe** [comment]
@Enkelmann  here is binaray.

[helloworld.zip](https://github.com/fkie-cad/cwe_checker/files/12492293/helloworld.zip)

PS: I cloned the latest cwe_checker source code, and build cwe_checker_installer, and run
we_checker --verbose helloworld

then got 
Error: missing field `register_calling_convention` at line 9330078 column 1

**Comment by drharryhe** [comment]
PSS: when i run analyzerHeaderless directly without postscript, namely:
./analyzeHeadless /tmp/cwe_checker  PcodeExtractor_helloworld_1693537411279  -import  helloworld 
it can analyze successfully.   So is it possible that something is wrong in the postScript?

**Comment by Enkelmann** [comment]
Thank you for the binary!

I was able to reproduce the issue on Linux, so it has nothing to do with Apple or ARM, but is probably just an ordinary bug in the cwe_checker. I am looking into it!

**Comment by drharryhe** [comment]
thanks for quick response. Could you please update me if any fix submitted.

**Comment by Enkelmann** [comment]
Current status: The original bug is fixed and the fix already merged to the master branch (PR #427). But I found another bug in the process (that also affects your `helloworld` binary), which I am currently trying to fix.

**Comment by Enkelmann** [comment]
The good news: Both bugs are fixed now on the master branch after PR #428 was merged today.

The bad news: Your binary is still a Mach-O binary and the cwe_checker does not support this binary format. Currently, we support ELF binaries and have some experimental support for PE binaries and bare-metal binaries. And it will probably stay this way for the foreseeable future, as all my development time right now goes into improving the analysis quality of the cwe_checker. I hope to someday also support Mach-O, but currently I cannot say when this might happen.

Closing this issue, since the bug fixes very likely also fixed the original issue reported here.
